src/machine/operation.c

```diff
@@ -0,0 +1,118 @@
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include "alloc-util.h"
+#include "fd-util.h"
+#include "operation.h"
+#include "process-util.h"
+
+static int operation_done(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        Operation *o = userdata;
+        int r;
+
+        assert(o);
+        assert(si);
+
+        log_debug("Operating " PID_FMT " is now complete with with code=%s status=%i",
+                  o->pid,
+                  sigchld_code_to_string(si->si_code), si->si_status);
+
+        o->pid = 0;
+
+        if (si->si_code != CLD_EXITED) {
+                r = sd_bus_error_setf(&error, SD_BUS_ERROR_FAILED, "Child died abnormally.");
+                goto fail;
+        }
+
+        if (si->si_status != EXIT_SUCCESS) {
+                if (read(o->errno_fd, &r, sizeof(r)) == sizeof(r))
+                        r = sd_bus_error_set_errnof(&error, r, "%m");
+                else
+                        r = sd_bus_error_setf(&error, SD_BUS_ERROR_FAILED, "Child failed.");
+
+                goto fail;
+        }
+
+        r = sd_bus_reply_method_return(o->message, NULL);
+        if (r < 0)
+                log_error_errno(r, "Failed to reply to message: %m");
+
+        operation_free(o);
+        return 0;
+
+fail:
+        r = sd_bus_reply_method_error(o->message, &error);
+        if (r < 0)
+                log_error_errno(r, "Failed to reply to message: %m");
+
+        operation_free(o);
+        return 0;
+}
+
+int operation_new(Manager *m, pid_t child, sd_bus_message *message, int errno_fd) {
+        Operation *o;
+        int r;
+
+        o = new0(Operation, 1);
+        if (!o)
+                return -ENOMEM;
+
+        r = sd_event_add_child(m->event, &o->event_source, child, WEXITED, operation_done, o);
+        if (r < 0) {
+                free(o);
+                return r;
+        }
+
+        o->pid = child;
+        o->message = sd_bus_message_ref(message);
+        o->errno_fd = errno_fd;
+
+        LIST_PREPEND(operations, m->operations, o);
+        m->n_operations++;
+        o->manager = m;
+
+        log_debug("Started new operation " PID_FMT ".", child);
+
+        /* At this point we took ownership of both the child and the errno file descriptor! */
+
+        return 0;
+}
+
+Operation *operation_free(Operation *o) {
+        if (!o)
+                return NULL;
+
+        sd_event_source_unref(o->event_source);
+
+        safe_close(o->errno_fd);
+
+        if (o->pid > 1)
+                (void) sigkill_wait(&o->pid);
+
+        sd_bus_message_unref(o->message);
+
+        if (o->manager) {
```

Comment:
It's a general coding convention we follow quite often in our sources: we want destructors to be safe to destruct even half-initialized objects, in case initialization failed half-way. This way we can use the same destructors for fully initialized objects and for incompletely initialized objects. (Also see CODING_STYLE about this). Now, if you have the case whre you have a set of "smaller" objects that are tracked by some singular "larger" object (i.e. in this case multiple "Operation" objects are tracked by a single "Manager" operation), then we usually link them up both ways at the same time, i.e. operation->manager is set to the manager object at the same time as the operation object is added to the manager's linked list. This way, the fact whether the "manager" ptr in the operation object is non-NULL may be used to determine whether the operation is in the linked list. Hence the check here: if o->manager is not NULL, then we remove the object from the linked list, as we know that the initialization of the operation object was already complete enough so that the objects where linked up.

Now, in this case, this all isn't strictly necessary, as our constructor for the operation object doesn't actually rely on the fact that the destrictor is written defensively like this, hence the exercise is indeed semi-pointless. However, I still think we should follow this scheme anyway, maybe one day the constructor for the operation object actually does want to use the destructor like this...

That all said, we only make use of this convention at many places of the code base, not in all...